src/v/cluster/security_manager.cc

```diff
@@ -153,8 +153,22 @@ security_manager::fill_snapshot(controller_snapshot& controller_snap) const {
 
     for (const auto& cred : _credentials.local()) {
         ss::visit(cred.second, [&](security::scram_credential scram) {
-            snapshot.user_credentials.push_back(user_and_credential{
-              security::credential_user{cred.first}, std::move(scram)});
+            if (scram.principal().has_value()) {
+                // Do not write ephemeral users to disk
+                if (
+                  scram.principal()->type()
+                  != security::principal_type::ephemeral_user) {
+                    snapshot.user_credentials.push_back(user_and_credential{
+                      security::credential_user{cred.first}, std::move(scram)});
+                }
+            } else {
+                // Regular users may not have a defined principal. For example,
+                // those users created from HTTP POST security/users (See
+                // parse_scram_credential() in the Admin server). Therefore, add
+                // these users into the snapshot.
+                snapshot.user_credentials.push_back(user_and_credential{
+                  security::credential_user{cred.first}, std::move(scram)});
+            }
```

Comment:
nitpick: I think it would be clearer to use a compound check at the top rather than repeating the insertion logic; the way it's written makes me stare it it for differences that aren't there.

Even if the formatter smashes it up a bit:

```c++
            bool is_not_ephemeral_user
              = !scram.principal().has_value()
                || scram.principal()->type()
                     != security::principal_type::ephemeral_user;
            if (is_not_ephemeral_user) {
                snapshot.user_credentials.push_back(user_and_credential{
                  security::credential_user{cred.first}, std::move(scram)});
            }
```